File: results/dnd-rest-benchmark/lifecycle-runs/20260730T082348Z_claude_claude-sonnet-5_go-stdlib/inventory.go
package main

import (
	"encoding/json"
	"net/http"
	"strings"
)

type inventoryEntry struct {
	ItemSlug string `json:"item_slug"`
	Quantity int    `json:"quantity"`
	Owner    string `json:"owner"`
}

type equipmentEntry struct {
	CharacterID string `json:"character_id"`
	ItemSlug    string `json:"item_slug"`
	Quantity    int    `json:"quantity"`
}

func inventoryEntryResponse(e inventoryEntry) map[string]interface{} {
	return map[string]interface{}{
		"item_slug": e.ItemSlug,
		"quantity":  e.Quantity,
		"owner":     e.Owner,
	}
}

func equipmentEntryResponse(e equipmentEntry) map[string]interface{} {
	return map[string]interface{}{
		"character_id": e.CharacterID,
		"item_slug":    e.ItemSlug,
		"quantity":     e.Quantity,
	}
}

func handleAddInventoryItem(w http.ResponseWriter, r *http.Request, campaignID string) {
	var req struct {
		ItemSlug string `json:"item_slug"`
		Quantity *int   `json:"quantity"`
		Owner    string `json:"owner"`
	}
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	if req.ItemSlug == "" || req.Owner == "" || req.Quantity == nil || *req.Quantity <= 0 {
		writeError(w, http.StatusBadRequest, "item_slug, quantity, and owner are required")
		return
	}

	campaignMu.Lock()
	c, exists := campaignStore[campaignID]
	if !exists {
		campaignMu.Unlock()
		writeError(w, http.StatusNotFound, "campaign not found")
		return
	}
	entry := inventoryEntry{ItemSlug: req.ItemSlug, Quantity: *req.Quantity, Owner: req.Owner}
	c.Inventory = append(c.Inventory, entry)
	campaignMu.Unlock()
	persistState()

	writeJSON(w, http.StatusCreated, inventoryEntryResponse(entry))
}

func handleAssignEquipment(w http.ResponseWriter, r *http.Request, campaignID, characterID string) {
	var req struct {
		ItemSlug string `json:"item_slug"`
		Quantity *int   `json:"quantity"`
	}
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	if req.ItemSlug == "" || req.Quantity == nil || *req.Quantity <= 0 {
		writeError(w, http.StatusBadRequest, "item_slug and quantity are required")
		return
	}

	campaignMu.Lock()
	c, exists := campaignStore[campaignID]
	if !exists {
		campaignMu.Unlock()
		writeError(w, http.StatusNotFound, "campaign not found")
		return
	}
	characterExists := false
	for _, ch := range c.Characters {
		if ch.ID == characterID {
			characterExists = true
			break
		}
	}
	if !characterExists {
		campaignMu.Unlock()
		writeError(w, http.StatusNotFound, "character not found")
		return
	}

	partyTotal := 0
	for _, e := range c.Inventory {
		if e.Owner == "party" && e.ItemSlug == req.ItemSlug {
			partyTotal += e.Quantity
		}
	}
	assignedTotal := 0
	for _, e := range c.Equipment {
		if e.ItemSlug == req.ItemSlug {
			assignedTotal += e.Quantity
		}
	}
	if partyTotal-assignedTotal < *req.Quantity {
		campaignMu.Unlock()
		writeError(w, http.StatusBadRequest, "insufficient party inventory for item")
		return
	}

	entry := equipmentEntry{CharacterID: characterID, ItemSlug: req.ItemSlug, Quantity: *req.Quantity}
	c.Equipment = append(c.Equipment, entry)
	campaignMu.Unlock()
	persistState()

	writeJSON(w, http.StatusOK, equipmentEntryResponse(entry))
}

func handleInventorySummary(w http.ResponseWriter, r *http.Request, campaignID string) {
	campaignMu.Lock()
	c, exists := campaignStore[campaignID]
	if !exists {
		campaignMu.Unlock()
		writeError(w, http.StatusNotFound, "campaign not found")
		return
	}
	partyItems := 0
	healingPotionsTotal := 0
	for _, e := range c.Inventory {
		if e.Owner == "party" {
			partyItems++
			if e.ItemSlug == "healing-potion" {
				healingPotionsTotal += e.Quantity
			}
		}
	}
	assignedItems := len(c.Equipment)
	healingPotionsAssigned := 0
	for _, e := range c.Equipment {
		if e.ItemSlug == "healing-potion" {
			healingPotionsAssigned += e.Quantity
		}
	}
	campaignMu.Unlock()

	writeJSON(w, http.StatusOK, map[string]interface{}{
		"campaign_id":               campaignID,
		"party_items":               partyItems,
		"assigned_items":            assignedItems,
		"healing_potions_available": healingPotionsTotal - healingPotionsAssigned,
	})
}

func handleCampaignInventorySub(w http.ResponseWriter, r *http.Request, campaignID, rest string) bool {
	if rest == "inventory" {
		if r.Method != http.MethodPost {
			writeError(w, http.StatusMethodNotAllowed, "method not allowed")
			return true
		}
		handleAddInventoryItem(w, r, campaignID)
		return true
	}
	if rest == "inventory/summary" {
		if r.Method != http.MethodGet {
			writeError(w, http.StatusMethodNotAllowed, "method not allowed")
			return true
		}
		handleInventorySummary(w, r, campaignID)
		return true
	}
	if strings.HasPrefix(rest, "characters/") && strings.HasSuffix(rest, "/equipment") {
		characterID := strings.TrimSuffix(strings.TrimPrefix(rest, "characters/"), "/equipment")
		if characterID == "" {
			return false
		}
		if r.Method != http.MethodPost {
			writeError(w, http.StatusMethodNotAllowed, "method not allowed")
			return true
		}
		handleAssignEquipment(w, r, campaignID, characterID)
		return true
	}
	return false
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260730T150315Z_pi_kimi-k2p7-code_typescript-node/src/routes/inventory.ts
import { ServerResponse } from 'node:http';
import { sendJSON } from '../http-utils.js';
import {
  addCharacterEquipment,
  addInventoryItem,
  campaignCharacterInCampaign,
  campaignExists,
  countEquipmentRows,
  countInventoryRows,
  decrementInventory,
  getPartyInventoryItem,
  sumInventoryQuantity,
} from '../repository.js';
import { isNonEmptyString, isPositiveInteger } from '../validators.js';

export function handleAddInventoryItem(
  res: ServerResponse,
  params: Record<string, string>,
  body: unknown,
): void {
  if (!campaignExists(params.id)) {
    sendJSON(res, 404, { error: 'not found' });
    return;
  }

  const { item_slug, quantity, owner } = body as Record<string, unknown>;
  if (!isNonEmptyString(item_slug) || !isPositiveInteger(quantity) || !isNonEmptyString(owner)) {
    sendJSON(res, 400, { error: 'invalid input' });
    return;
  }

  addInventoryItem(params.id, item_slug, quantity, owner);
  sendJSON(res, 201, { item_slug, quantity, owner });
}

export function handleAssignEquipment(
  res: ServerResponse,
  params: Record<string, string>,
  body: unknown,
): void {
  if (!campaignExists(params.id)) {
    sendJSON(res, 404, { error: 'not found' });
    return;
  }
  if (!campaignCharacterInCampaign(params.character_id, params.id)) {
    sendJSON(res, 404, { error: 'not found' });
    return;
  }

  const { item_slug, quantity } = body as Record<string, unknown>;
  if (!isNonEmptyString(item_slug) || !isPositiveInteger(quantity)) {
    sendJSON(res, 400, { error: 'invalid input' });
    return;
  }

  const partyItem = getPartyInventoryItem(params.id, item_slug);
  if (!partyItem || partyItem.quantity < quantity) {
    sendJSON(res, 400, { error: 'insufficient quantity' });
    return;
  }

  decrementInventory(params.id, item_slug, 'party', quantity);
  addCharacterEquipment(params.id, params.character_id, item_slug, quantity);
  sendJSON(res, 200, { character_id: params.character_id, item_slug, quantity });
}

export function handleInventorySummary(
  res: ServerResponse,
  params: Record<string, string>,
): void {
  if (!campaignExists(params.id)) {
    sendJSON(res, 404, { error: 'not found' });
    return;
  }

  sendJSON(res, 200, {
    campaign_id: params.id,
    party_items: countInventoryRows(params.id),
    assigned_items: countEquipmentRows(params.id),
    healing_potions_available: sumInventoryQuantity(params.id, 'healing-potion', 'party'),
  });
}
